lib/Connection.php

```diff
@@ -109,9 +109,18 @@ public function pushMessage(Message $message, ?bool $masked = null): void
         ]);
     }
 
+    private function waitForMessage(): void
+    {
+        // non-blocking way to wait for stream to change
+        $read = [$this->stream];
+        stream_select($read, null, null, $this->options['timeout'] ?? PHP_INT_MAX, 0);
+    }
+
     // Pull a message from stream
     public function pullMessage(): Message
     {
+        $this->waitForMessage();
```

Comment:
This should be right before `fread()` in `read()`, since `read()` is called multiple times as part of `pullMessage()`, and consecutive reads would also block.